sympy/polys/powerseriesring.py

```diff
@@ -0,0 +1,241 @@
+from sympy.polys.domains import Domain
+from sympy.utilities import public
+
+
+ZZSeries: TypeAlias = Any
+QQSeries: TypeAlias = Any
+
+
+def series_from_list(series: Any, prec: int | None = None) -> str:
+    terms = []
+
+    for i, coeff in enumerate(series):
+        if coeff == 0:
+            continue
+        if coeff == 1 and i != 0:
+            term = "x" if i == 1 else f"x**{i}"
+        elif coeff == -1 and i != 0:
+            term = "-x" if i == 1 else f"-x**{i}"
+        else:
+            if i == 0:
+                term = f"{coeff}"
+            elif i == 1:
+                term = f"{coeff}*x"
+            else:
+                term = f"{coeff}*x**{i}"
+        terms.append((i, term))
+
+    if not terms:
+        return "0"
+
+    # Sort terms based on exponent depending on `prec`
+    terms.sort(reverse=(prec is None))  # Descending if prec is None
+    poly = " + ".join(term for _, term in terms).replace("+ -", "- ")
+
+    if prec is not None:
+        return f"{poly} + O(x**{prec})"
+    return poly
+
+@public
+def power_series_ring(
+    domain: Domain, prec: int = 6
+) -> PowerSeriesRingZZ | PowerSeriesRingQQ:
+    """
+    Create a power series ring over the given domain.
+
+    Parameters
+    ----------
+    domain : Domain
+        The ground domain for the power series ring. Must be ZZ or QQ.
+    prec : int, optional
+        The default precision for power series operations. Default is 6.
+
+    Returns
+    -------
+    PowerSeriesRingZZ | PowerSeriesRingQQ
+        A power series ring instance. The actual implementation (Flint or Python)
+        depends on whether python-flint is available.
+
+    Examples
+    --------
+    >>> from sympy.polys.domains import ZZ, QQ
+    >>> from sympy.polys.powerseriesring import power_series_ring
+    >>> R_ZZ = power_series_ring(ZZ, 5)
+    >>> R_QQ = power_series_ring(QQ, 10)
+    """
+    flint: bool | None = None
+    if GROUND_TYPES == 'flint':
+        flint = True
+
+    if domain.is_ZZ:
+        if flint:
+            from sympy.polys.flint_powerseriesring import FlintPowerSeriesRingZZ
+            return FlintPowerSeriesRingZZ(prec)
+        from sympy.polys.python_powerseriesring import PythonPowerSeriesRingZZ
+        return PythonPowerSeriesRingZZ(prec)
+    if domain.is_QQ:
+        if flint:
+            from sympy.polys.flint_powerseriesring import FlintPowerSeriesRingQQ
+            return FlintPowerSeriesRingQQ(prec)
+        from sympy.polys.python_powerseriesring import PythonPowerSeriesRingQQ
+        return PythonPowerSeriesRingQQ(prec)
+    raise TypeError("Ground domain must be an instance of QQ or ZZ")
+
+
+@runtime_checkable
+class PowerSeriesRingZZ(Protocol):
+    """Protocol for power series ring over integer ring.
+
+    Attributes
+    ----------
+    domain : Domain
+        The ground domain (ZZ).
+    prec : int
+        The default precision for power series operations.
+    """
+
+    domain: Domain
+    prec: int
```

Comment:
It is better in the protocol to define these as `@property` methods so that the type checker treats them as read-only.